[11.0] [ADD] web_timeline: Dependency arrows

tarteo opened the pull request:
Adds a new optional attribute dependency_arrow.

If the attribute is set it will look like:
![image](https://user-images.githubusercontent.com/10028499/44399992-e05a2580-a549-11e8-9799-bf0ffa7aa501.png)


I created a widget TimelineCanvas which does all the svg stuff. This way the renderer will not be bloated with it.

tarteo committed:
[ADD] web_timeline: New dependency_arrow attribute

Update README.rst

[FIX] Remove console.log

[ADD] Make timeline.fit optional

[FIX] Use stringified points

[IMP] Reversed the arrow head and fixed lint issues

[IMP] Use options parameter for line color and width

[FIX] Version number

[IMP] Minor improvements

pedrobaeza commented:
Another great addition! With this and the planning module we are doing, we finally have a superb project app!

astirpe reviewed (changes requested):
Could you increase the version in manifest? Also in https://github.com/OCA/project/pull/420 should be stated that the minimum version of the required web_timeline module is the one you are going to set.

tarteo commented:
I've put the arrows UML way, but I think it's more logical for a timeline which is mostly read left to right, to put the arrowhead on the depending side. What do you think?

pedrobaeza commented:
When I first saw it, I think there was something weird on it, but being logical when I thought about it. I'm with your that maybe we should reverse it for the shake of clarity.

astirpe commented:
@tarteo yes should be reversed

yajo commented on web_timeline/static/src/js/timeline_controller.js:
```diff
@@ -226,13 +229,17 @@ var CalendarController = AbstractController.extend({
         };
-        this.update(params, null);
+
+        var defaults = _.defaults(options || {}, {
```
`_.defaults({}, options, {` please. This way we avoid modifying the `options` object in place.

yajo reviewed (changes requested):
Looks pretty nice! Just a little thing.

Also please [review the JS linter](https://travis-ci.org/OCA/web/jobs/417320520#L769-L806), which doesn't put travis in :x:  currently, but should be honored if possible (and if it has errors, it should be reported, so we polish that linter too :blush: ).

Thanks!

simahawk commented on web_timeline/static/src/js/timeline_renderer.js:
```diff
@@ -176,9 +178,41 @@ var CalendarRenderer = AbstractRenderer.extend({
+        if (!from.displayed || !to.displayed)
+            return;
+
+        this.canvas.draw_arrow(from.dom.box, to.dom.box, 'black', 1);
```
shouldn't "black" come from an option?

tarteo commented:
@pedrobaeza @astirpe  I've upped the version number and reversed the arrow head. You can see the result in the PR description.
@Jayo I've applied your suggestion and cleaned up the lint (I couldn't get rid of some of the errors) I thought about using directives but I rather let it print the error instead of hiding it. It's not causing the PR to be red so it's not a problem.
@simahawk I've changed the function so the line color and line width are changeable with the options parameter.

simahawk reviewed (approved):
LG

pedrobaeza commented:
@Yajo waiting your blessing for merging this

pedrobaeza commented:
@tarteo I think the next step in the roadmap should be the possibility of defining dependencies graphically, having for example a dot in timeline object from which you can drag and drop to other element's dot for setting the dependency. How do you see it?

yajo commented on web_timeline/static/src/js/timeline_renderer.js:
```diff
@@ -314,16 +330,9 @@ var CalendarRenderer = AbstractRenderer.extend({
         });
 
-        var content = evt.__name != undefined ? evt.__name : evt.display_name;
+        var content = typeof evt.__name === 'undefined' ? evt.display_name : evt.__name;
```
I recommend you to use `!_.isUndefined(evt)` instead.

yajo commented on web_timeline/static/src/js/timeline_renderer.js:
```diff
@@ -342,10 +351,25 @@ var CalendarRenderer = AbstractRenderer.extend({
     },
 
+    render_timeline_item: function(evt) {
+        if(this.qweb.has_template('timeline-item')) {
```
Since `this.qweb` is conditionally declared, this could raise the common `TypeError: undefined has no property "has_template"` error.

It would be better IMHO to always have a `this.qweb` element, that may or may not have templates.

tarteo replied:
Made this.qweb always available even when there are no templates

yajo commented on web_timeline/static/src/js/timeline_renderer.js:
```diff
@@ -342,10 +351,25 @@ var CalendarRenderer = AbstractRenderer.extend({
         // handle a click on a group header
-        if (e.what === 'group-label' && e.group != -1) {
-            this._trigger(e, function(){}, 'onGroupClick');
+        if (e.what === 'group-label' && e.group !== -1) {
```
You can use `~e.group` instead of `e.group !== -1` :wink:

yajo reviewed (approved):
Good!

tarteo commented:
@pedrobaeza Yes, for the dependencies that's exactly how I see it. E.g. It would also be nice to use parent tasks to print a kind of collapse / expand control.
